HW2/P4/driver.py

```diff
@@ -15,15 +15,60 @@
 from timer import Timer
 import threading
 
+def filter_component(in_image, out_image, events, num_iterations, num_threads):
+    # Get the current thread id so we know what rows to operate on
+    curr_thread_id = int(threading.currentThread().getName())
+
+    #print 'Current thread is:', curr_thread_id
+
+    # Now filter as many times as we desire
+    for ii in range(num_iterations):
+        # Make sure that the necessary computations are already completed
+        # Clearly on the first iteration we should just go ahead
+        # The modular arithmetic just wraps around for edge cases
+        if ii > 0:
+            events[curr_thread_id][ii - 1].wait()
+            events[(curr_thread_id - 1) % num_threads][ii - 1].wait()
+            events[(curr_thread_id + 1) % num_threads][ii - 1].wait()
+
+        # Now that we've waited for everything, we can definitely go ahead
+        #print 'About to filter on iteration', ii
+        filtering.median_3x3(in_image, out_image, curr_thread_id, num_threads)
+        #print 'After filter on iteration', ii
+
+        # Swap the direction of filtering
+        in_image, out_image = out_image, in_image
+
+        # And let everybody else know that we finished up
+        events[curr_thread_id][ii].set()
+
+
+
 def py_median_3x3(image, iterations=10, num_threads=1):
     ''' repeatedly filter with a 3x3 median '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
 
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
+    # Create a matrix of events corresponding to (thread, iteration) pairs
+    # Essentially, we will set event (t, i) when thread t has completed iteration i
+    # Then when thread thread t' wants to try iteration i', we see if (t', i'-1), (t'-1, i-1), (t'+1, i-1) are done
+    events  = [[threading.Event() for ii in range(iterations)] for jj in range(num_threads)]
+
```

Comment:
Nice! I did the same thing as you here, creating an Event for each thread, iteration pair.